src/shinylive.ts

```diff
@@ -0,0 +1,512 @@
+    // create directory based on uri
+    await vscode.workspace.fs.createDirectory(uri);
+  }
+
+  lastUsedDir = path.dirname(uri.path);
+  return uri;
+}
+
+/**
+ * Encode a Shinylive bundle into a URL string.
+ *
+ * @param {ShinyliveBundle} { language, files, mode } A Shinylive bundle object
+ * with the language, files, and mode to encode.
+ * @returns {string} The encoded Shinylive URL.
+ */
+function shinyliveUrlEncode({ language, files, mode }: ShinyliveBundle) {
+  const filesJson = JSON.stringify(files);
+  const filesLZ = lzstring.compressToEncodedURIComponent(filesJson);
+
+  const host = vscode.workspace
+    .getConfiguration("shiny.shinylive")
+    .get<string>("host");
+
+  let includeHeader;
+  if (mode === "app") {
+    // Header is only relevant in the `app` mode
+    includeHeader = vscode.workspace
+      .getConfiguration("shiny.shinylive")
+      .get<boolean>("includeHeader");
+  }
+  const h = includeHeader ? "" : "h=0&";
+
+  return `${host}/${language}/${mode}/#${h}code=${filesLZ}`;
+}
+
+/**
+ * Decode a Shinylive URL string into a Shinylive bundle object.
+ *
+ * @param {string} url The Shinylive URL string to decode.
+ * @returns {ShinyliveBundle | undefined} The decoded Shinylive bundle, or
+ * `undefined` if the URL could not be decoded.
+ */
+function shinyliveUrlDecode(url: string): ShinyliveBundle | undefined {
+  const { hash, pathname } = new URL(url);
+  const { searchParams } = new URL(
+    "https://shinylive.io/?" + hash.substring(1)
+  );
+
+  const code = searchParams.get("code");
+
+  if (!code) {
+    return;
+  }
+
+  const filesJson = lzstring.decompressFromEncodedURIComponent(code);
+  const files = JSON.parse(filesJson);
+
+  const pathParts = pathname.split("/");
+
+  return {
+    language: pathParts[1] as ShinyliveLanguage,
+    files: files as ShinyliveFile[],
+    mode: pathParts[2] as ShinyliveMode,
+  };
+}
+
+/**
+ * Write a list of Shinylive files to the output directory.
+ *
+ * @async
+ * @param {ShinyliveFile[]} files A list of files to write to the output
+ * directory.
+ * @param {vscode.Uri} outputDir The directory where the files will be saved.
+ * @returns {Promise<string[]>} A list of the local file paths where the files
+ * were saved.
+ */
+async function shinyliveWriteFiles(
+  files: ShinyliveFile[],
+  outputDir: vscode.Uri
+): Promise<string[]> {
+  const localFiles = [];
+
+  for (const file of files) {
+    const filePath = vscode.Uri.file(`${outputDir.fsPath}/${file.name}`);
+    const fileDir = vscode.Uri.file(path.dirname(filePath.fsPath));
+
+    if (!fs.existsSync(fileDir.fsPath)) {
+      await vscode.workspace.fs.createDirectory(fileDir);
+    }
+
+    let contentBuffer: Buffer;
+
+    switch (file.type) {
+      case "binary":
+        contentBuffer = Buffer.from(file.content, "base64");
+        break;
+      default:
+        contentBuffer = Buffer.from(file.content);
+        break;
+    }
+
+    await vscode.workspace.fs.writeFile(filePath, contentBuffer);
```

Comment:
Worth a confirmation prompt to overwrite existing files?

Reply:
For single-file Shinylive apps, the `showSaveDialog()` window will ask users to confirm if they want to overwrite an existing file.

For multi-file Shinylive apps, `showSaveDialog()` forces users to pick a non-existent directory (it won't let you move forward with an existing directory).

That said, if we tried to re-use this code in a different context, we'd probably appreciate the back-up confirmation, but we'd have to do disable those checks for our current code to avoid confirming with the user more than once.

This whole line of thought does bring up that because we give the files in the bundle paths relative to the primary app file, we could end up with files that want to be saved in a parent directory. We should probably do something in this case -- I'm leaning toward throwing an error rather than trying to work it out or let the extension write to arbitrary directories.

Reply:
I update the extension so that it...

1. Throws early when given a Shinylive link that contains files that would be added outside the selected directory.
    * Here's [an example app that should fail to save](https://shinylive.io/r/editor/#h=0&code=NobwRAdghgtgpmAXGKAHVA6ASmANGAYwHsIAXOMpMAGwEsAjAJykYE8AKAZwAtaJWAlAB0IIgMQACACJwAZnzgSAqgEkJsoowkATKKSic4pCQDdacAO5wtaVBIC0j+yICutCQB5766m+0AFKABzOHYREQkJSQBBdAlSWlJqRSdnCEiEpLhAiDhqMLAsOCgCBLNSViEwAVxw9KiJAGVabTh6FglqKFYiF2MLRO4JPlQ+iSgIbQle0lHjVvkIRNoSTgcnCIlOFraWABlumbDResjJZtb2rVQJvPVNYYg5tdTNyO3Llhy847fIhpUTz6iAkABU4AAPYwaa6MIhmbR8ILjCQENAJEjrRx-M4SAByRHIIIAwtwJiE1jAoK14kR4txFGjUBj0nx6YpyFDAXNUSRSHDqDiGixFC5ULpyFM2aQGdM+jyRVBhjB4Ii9HlWOM1j0XPFWKg4ELOaRuX12CM+iopgBeCRVJksqq1U7-V1uzpQeh3W1VYnolYQRBOoXu10mKC+RQ+sBSPRKxouFUsSrVZ1CySm0ggxp5OClB4w1HcIhEbYQZESgxGLFpd2GZKlTPmoEmm12sCVwykYMu0Ohrpe6gSaOkkuGFGdoxBvAhvv-AjF2gEOBrW0EApwggAayd7dQjBXnBcB939pYnCqAhqdXdGZbILxiesS4kFH5msLEETXq0RFk03oNZSDpMxLBrIUv3gRglybC1W2HdsiEAns51QgdvXbR8YB-aZ-yQwxGHDFkgJA8wLGnZ1ULncNIwQgBGAAGYQTlda8WP+SQAFkoDZG5ciHQtEU4VAulYJE5VmPoXg2F0qT4b58hvN1JAAeXlYEJAAMU0KlSEleJIWhB4HQDcDewkbgAGZ2GNNTJNIAoTJIXdiDIHjci0W1hImK803M1T1KzCQADVrHaBIYAMqF7i0SstkTKk2DM90TDCvRaBgcEoTsuYCiPJM2EvPzbwkHKNIACVBTi9niT1kgkAYZQkA8AEcXBXfTIJwv8AIIoiA2k7FzP0ehkjKhyqlAixLyUyJmLmkRmPEaQ5AULZrFSrRqCIIIX2A+KCtoAAvRQJimKb1obfTJ2MV4ID66xPG8WQXAgUoA2bOZcAkuYBAkEAlMkIpSGPdIZUUVr2s4a64y7ZKGgAIU1VoCC6aDyx0WGjEzLUUQPEoykUSF90PUyrFfe7j0NepJBlPQgxpzZJDojAJBUYxaDWEhqE1NFqGSKYLAZMHZTgtZEh0OADUmLn0gXckVyZiQACZWdBWViBgOZ0sxM7mpXFxqGMEUtjJA8pnoTUIyHcHUWt6xOEohp-loDA4FZiXuc1SE8z6FcUTLIJ6oi6nIhunGvH1gnaFS9gAZdThGoXT6+gAEhup25yqTcdzABCc8zvsqhJzgjxPPPbRLsu4EL0Mz0YC8K7thv5okABfJaaYkYl8fIO3mVMu7cXVxQZjmVOnNZNZNbmOALarKZdaj97UtfCEq4DJW6eMA8QcYe7HnHyfWYAdWF9kJBcAiiwVoCGSVs8B+cvP5DybQGbOJWWbBXg1het6WTDDWFAPoRBdJLmtrze2Ux9oHhnn7C+Y8+hK1VviMCk8gESFGDwOeEh2jblpBfJgRALDX0LAeewQkRLdCZkrAkfdt54LzCA8cttKz2E0OYMguD8YryJuvA8pcBpK0HCQnQJAADk8wpYUEXqyFsE9-QkG+jKUs-CN6rHGCeLueIVKgjtvzXBQsKCHzThg+W5ZFb1CQaQRRT90iRwPJMawWUHLx1dHBOxjp6gd0BhIAA4hQaw6oA4JWTLhC+N14aSBHj9NO+VErIxkTLaYItFDhxbMvQmStiaCO2MorYdIBj8yYfrewPsCB+0FsLOAm1Mb6C7DjTmSs+A0RaOqbQ31XbuwajUupttPFRIsRSTYNjU4JPCY42R1h-DozcX8KJkcMm5Vbu8MJbB2A3Vbr49i5xiwWAvvIBuxgqiiDzvhDaOsD5DwaLEsZF1WjS20LLPBhIhhsLjO0QwONeHZK7rk0mutJimNsfhb6nA6QS2KUOL0ZSKlVN6UEzaLSFFRIeJ4-CmDhlz1GYFVOF0pnOMYKCOqoR3H-AZNSTZWMTQtnYDUCQ6RbQYsAtszubc6iSB7sUPujReD8HGHEO6PA+CsFiKgdgbhwUbWsMxPAYAKgGioMaMAbcAC6QA)
1. Confirms before overwriting files, except for single-file Shinylive links because VS Code will already have confirmed.

Reply:
Yikes, Shinylive probably shouldn't let you create files like that...

Reply:
If we don't let people pick existing directories, great! In the future if the overwrite confirmation actually gets triggered, it'd be really annoying to have to say yes repeatedly, one overwrite confirmation would be better. But I'm perfectly fine merging this as-is now.